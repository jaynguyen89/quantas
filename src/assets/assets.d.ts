declare module '*.png' {
    const path: string;
    export default path;
}

declare module 'blob-polyfill';
