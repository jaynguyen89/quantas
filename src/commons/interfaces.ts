import { CancellationType, ImageType, PromotionType, RatingType } from 'src/commons/enums';

export interface IApiHotel {
    id: string,
    property: {
        propertyId: string,
        title: string,
        address: Array<string>,
        previewImage: {
            url: string,
            caption: string,
            imageType: string,
        },
        rating: {
            ratingValue: number,
            ratingType: string,
        },
    },
    offer: {
        promotion: {
            title: string,
            type: string,
        },
        name: string,
        displayPrice: {
            amount: number,
            currency: string,
        },
        savings: {
            amount: number,
            currency: string,
        },
        cancellationOption: {
            cancellationType: string,
        },
    },
}

export interface IHotel {
    id: string,
    name: string,
    address: string,
    image: {
        url: string,
        caption: string,
        type: ImageType,
    },
    rating: {
        value: number,
        type: RatingType,
    },
    pricing: IOffer,
}

export interface IOffer {
    name: string,
    price: {
        value: number,
        currency: string,
    },
    promotion: {
        title: string,
        type: PromotionType,
    },
    savings?: {
        value: number,
        currency: string,
    },
    cancellationType: CancellationType,
}