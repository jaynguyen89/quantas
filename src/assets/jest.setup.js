import { TextEncoder, TextDecoder, ReadableStream } from 'node:util';

global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;
global.ReadableStream = ReadableStream;