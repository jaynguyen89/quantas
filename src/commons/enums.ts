export enum ImageType {
    PRIMARY = 'PRIMARY',
    SECONDARY = 'SECONDARY',
}

export enum PromotionType {
    MEMBER = 'MEMBER',
    GUEST = 'GUEST',
}

export enum CancellationType {
    FREE_CANCELLATION = 'FREE_CANCELLATION',
    NOT_REFUNDABLE = 'NOT_REFUNDABLE',
}

export enum RatingType {
    SELF = 'SELF',
    STAR = 'STAR',
}

export enum SortingOption {
    HIGH_LOW = 'Price high-low',
    LOW_HIGH = 'Price low-high',
}