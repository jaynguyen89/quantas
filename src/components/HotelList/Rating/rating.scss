@use 'src/commons/variables' as vars;

.rating {
    display: inline-block;
    unicode-bidi: bidi-override;
    color: vars.$grey;
    font-size: vars.$h9;
    margin: 0 0 0 vars.$sh9;
    width: auto;
    position: relative;
    padding: 0;

    .front {
        color: vars.$yellow;
        padding: 0;
        position: absolute;
        z-index: 1;
        display: flex;
        top: 0;
        left: 0;
        overflow: hidden;

        .circle {
            width: vars.$xh10;
            height: vars.$xh10;
            margin: 0 vars.$xmicro;
            border-radius: 50%;
            background-color: vars.$yellow;
        }
    }

    .behind {
        padding: 0;
        display: flex;
        z-index: 0;

        .circle {
            width: vars.$xh10;
            height: vars.$xh10;
            margin: 0 vars.$xmicro;
            border-radius: 50%;
            background-color: vars.$grey;
        }
    }

    @media screen and (max-width: vars.$desktop-width) {
        display: block;
        margin: vars.$sh10 0;
    }
}