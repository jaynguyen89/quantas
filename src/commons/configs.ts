import * as process from 'process';
import { SortingOption } from 'src/commons/enums';

const configs = {
    maxRatingCap: process.env.REACT_APP_MAX_RATING_CAP === undefined
        ? 5
        : +process.env.REACT_APP_MAX_RATING_CAP,
    defaultSortOption: (process.env.REACT_APP_DEFAULT_SORTING ?? 'HIGH_LOW') as keyof typeof SortingOption,
};

export default configs;