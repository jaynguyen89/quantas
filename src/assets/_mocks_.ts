import { IApiHotel, IHotel } from 'src/commons/interfaces';
import { CancellationType, ImageType, PromotionType, RatingType } from 'src/commons/enums';

export const mockedApiHotelData: {
    results: Array<IApiHotel>,
} = {
    results: [
        {
            id: '123',
            property: {
                propertyId: '111',
                title: 'First Hotel',
                address: ['1 What Street', 'Where 3333'],
                previewImage: {
                    url: 'https://photo1.png',
                    caption: 'Photo1',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 3,
                    ratingType: 'self',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo1',
                    type: 'MEMBER',
                },
                name: 'Promo1',
                displayPrice: {
                    amount: 111,
                    currency: 'AU',
                },
                savings: {
                    amount: 11,
                    currency: 'AU',
                },
                cancellationOption: {
                    cancellationType: 'NOT_REFUNDABLE',
                },
            },
        },
        {
            id: '456',
            property: {
                propertyId: '222',
                title: 'Second Hotel',
                address: ['2 Which Way', 'Anywhere 2222'],
                previewImage: {
                    url: 'https://photo2.png',
                    caption: 'Photo2',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 4,
                    ratingType: 'star',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo2',
                    type: 'MEMBER',
                },
                name: 'Promo2',
                displayPrice: {
                    amount: 222,
                    currency: 'US',
                },
                savings: null,
                cancellationOption: {
                    cancellationType: 'FREE_CANCELLATION',
                },
            },
        },
        {
            id: '789',
            property: {
                propertyId: '333',
                title: 'Third Hotel',
                address: ['3 Where Road', 'Somewhere 1111', 'Australia'],
                previewImage: {
                    url: 'https://photo3.png',
                    caption: 'Photo3',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 5,
                    ratingType: 'self',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo3',
                    type: 'MEMBER',
                },
                name: 'Promo3',
                displayPrice: {
                    amount: 333,
                    currency: 'VND',
                },
                savings: {
                    amount: 33,
                    currency: 'VND',
                },
                cancellationOption: {
                    cancellationType: 'FREE_CANCELLATION',
                },
            },
        },
    ],
};

export const mockedHotelData: Array<IHotel> = [
    {
        id: '123',
        name: 'First Hotel',
        address: '1 What Street, Where 3333',
        image: {
            url: 'https://photo1.png',
            caption: 'Photo1',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 3,
            type: RatingType.SELF,
        },
        pricing: {
            name: 'Promo1',
            price: {
                value: 111,
                currency: 'AU',
            },
            promotion: {
                title: 'Promo1',
                type: PromotionType.MEMBER,
            },
            savings: {
                value: 11,
                currency: 'AU',
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
    {
        id: '456',
        name: 'Second Hotel',
        address: '2 Which Way, Where 2222',
        image: {
            url: 'https://photo2.png',
            caption: 'Photo2',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 4,
            type: RatingType.STAR,
        },
        pricing: {
            name: 'Promo2',
            price: {
                value: 222,
                currency: 'US',
            },
            promotion: {
                title: 'Promo2',
                type: PromotionType.MEMBER,
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
    {
        id: '789',
        name: 'Third Hotel',
        address: '3 Where Road, Somewhere 1111, Australia',
        image: {
            url: 'https://photo3.png',
            caption: 'Photo3',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 5,
            type: RatingType.SELF,
        },
        pricing: {
            name: 'Promo3',
            price: {
                value: 333,
                currency: 'US',
            },
            promotion: {
                title: 'Promo3',
                type: PromotionType.MEMBER,
            },
            savings: {
                value: 33,
                currency: 'US',
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
];