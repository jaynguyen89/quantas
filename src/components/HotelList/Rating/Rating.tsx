import React, { JSX, useMemo } from 'react';
import 'src/components/HotelList/Rating/rating.scss';
import { RatingType } from 'src/commons/enums';
import configs from 'src/commons/configs';

type TRatingProps = {
    value: number,
    type: RatingType,
};

const Rating = ({
    value,
    type,
}: TRatingProps) => {
    const ratingIcon: JSX.Element = useMemo(
        () => type === RatingType.SELF
            ? <div className='circle' />
            : <span>â˜…</span>,
        [type],
    );

    return (
        <div className='rating'>
            <div className='front' style={{ width: `${value}%` }}>
                {[...Array(configs.maxRatingCap)].map((_, i) => (
                    <div id='icon-front' key={i}>{ratingIcon}</div>
                ))}
            </div>
            <div className='behind'>
                {[...Array(configs.maxRatingCap)].map((_, i) => (
                    <div id='icon-behind' key={i}>{ratingIcon}</div>
                ))}
            </div>
        </div>
    );
};

export default Rating;