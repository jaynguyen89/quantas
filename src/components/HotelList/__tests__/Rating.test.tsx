import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import toJson from 'enzyme-to-json';
import { render } from '@testing-library/react';
import { mockedHotelData } from 'src/assets/_mocks_';
import configs from 'src/commons/configs';
import Rating from 'src/components/HotelList/Rating/Rating';
import { RatingType } from 'src/commons/enums';

configure({ adapter: new Adapter() });

describe('Rating.tsx', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<Rating {...mockedHotelData[0].rating} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders ratings with circles', () => {
        const wrapper = render(<Rating {...mockedHotelData[0].rating} />);

        const frontEl = wrapper.container.getElementsByClassName('front');
        expect(frontEl.length).toBe(1);
        expect(frontEl[0]).toBeDefined();

        const circleFrontEl = frontEl[0].getElementsByTagName('div');
        expect(circleFrontEl.length).toBe(configs.maxRatingCap * 2);

        [...Array(configs.maxRatingCap)].forEach((_, i) => {
            if (circleFrontEl[i].hasAttribute('id'))
                expect(circleFrontEl[i].innerHTML).toEqual('<div class=\"circle\"></div>');
        });

        const behindEl = wrapper.container.getElementsByClassName('behind');
        expect(behindEl.length).toBe(1);
        expect(behindEl[0]).toBeDefined();

        const circleBehindEl = behindEl[0].getElementsByTagName('div');
        expect(circleBehindEl.length).toBe(configs.maxRatingCap * 2);

        [...Array(configs.maxRatingCap)].forEach((_, i) => {
            if (circleBehindEl[i].hasAttribute('id'))
                expect(circleBehindEl[i].innerHTML).toEqual('<div class=\"circle\"></div>');
        });
    });

    it('renders ratings with stars', () => {
        const data = mockedHotelData[0].rating;
        data.type = RatingType.STAR;
        const wrapper = render(<Rating {...data} />);

        const frontEl = wrapper.container.getElementsByClassName('front');
        expect(frontEl.length).toBe(1);
        expect(frontEl[0]).toBeDefined();

        const circleFrontEl = frontEl[0].getElementsByTagName('div');
        expect(circleFrontEl.length).toBe(configs.maxRatingCap);

        [...Array(configs.maxRatingCap)].forEach((_, i) => {
            if (circleFrontEl[i].hasAttribute('id'))
                expect(circleFrontEl[i].innerHTML).toEqual('<span>★</span>');
        });

        const behindEl = wrapper.container.getElementsByClassName('behind');
        expect(behindEl.length).toBe(1);
        expect(behindEl[0]).toBeDefined();

        const circleBehindEl = behindEl[0].getElementsByTagName('div');
        expect(circleBehindEl.length).toBe(configs.maxRatingCap);

        [...Array(configs.maxRatingCap)].forEach((_, i) => {
            if (circleBehindEl[i].hasAttribute('id'))
                expect(circleBehindEl[i].innerHTML).toEqual('<span>★</span>');
        });
    });
});