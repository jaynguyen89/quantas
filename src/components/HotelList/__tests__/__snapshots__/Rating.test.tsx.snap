// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating.tsx renders correctly 1`] = `
<div
  className="rating"
>
  <div
    className="front"
    style={
      Object {
        "width": "3%",
      }
    }
  >
    <div
      id="icon-front"
      key="0"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-front"
      key="1"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-front"
      key="2"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-front"
      key="3"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-front"
      key="4"
    >
      <div
        className="circle"
      />
    </div>
  </div>
  <div
    className="behind"
  >
    <div
      id="icon-behind"
      key="0"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-behind"
      key="1"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-behind"
      key="2"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-behind"
      key="3"
    >
      <div
        className="circle"
      />
    </div>
    <div
      id="icon-behind"
      key="4"
    >
      <div
        className="circle"
      />
    </div>
  </div>
</div>
`;
